lib/cockpit-components-listing.jsx

```diff
@@ -17,257 +17,267 @@
-                                 // contains tab indices
-            };
-        },
-        handleNavigateClick: function(e) {
-            // only consider primary mouse button
-            if (!e || e.button !== 0)
-                return;
-            this.props.navigateToItem();
-        },
-        handleExpandClick: function(e) {
-            // only consider primary mouse button
-            if (!e || e.button !== 0)
-                return;
+var React = require('react');
 
-            var willBeExpanded = !this.state.expanded;
-            this.setState( { expanded: willBeExpanded });
+require('./cockpit-components-listing.css');
 
-            var loadedTabs = {};
-            // unload all tabs if not expanded
-            if (willBeExpanded) {
-                // see if we should preload some tabs
-                var tabIdx;
-                var tabPresence;
-                for (tabIdx = 0; tabIdx < this.props.tabRenderers.length; tabIdx++) {
-                    if ('presence' in this.props.tabRenderers[tabIdx])
-                        tabPresence = this.props.tabRenderers[tabIdx].presence;
-                    else
-                        tabPresence = 'default';
-                    // the active tab is covered by separate logic
-                    if (tabPresence == 'always')
-                        loadedTabs[tabIdx] = true;
-                }
-                // ensure the active tab is loaded
-                loadedTabs[this.state.activeTab] = true;
-            }
+/* entry for an alert in the listing, can be expanded (with details) or standard
+ * columns list of columns to show in the header
+ *     columns to show, can be a string, react component or object with { name: 'name', 'header': false }
+ *     'header' (or if simple string) defaults to false
+ *     in case 'header' is true, <th> is used for the entries, otherwise <td>
+ * tabRenderers optional: list of tab renderers for inline expansion, array of objects with
+ *     - name tab name (has to be unique in the entry, used as react key)
+ *     - renderer react component
+ *     - data render data passed to the tab renderer
+ *     - presence 'always', 'onlyActive', 'loadOnDemand', default: 'loadOnDemand'
+ *         - 'always' once a row is expanded, this tab is always rendered, but invisible if not active
+ *         - 'onlyActive' the tab is only rendered when active
+ *         - 'loadOnDemand' the tab is first rendered when it becomes active, then follows 'always' behavior
+ *     if tabRenderers isn't set, item can't be expanded inline
+ * navigateToItem optional: callback triggered when a row is clicked, pattern suggests navigation
+ *     to view expanded item details, if not set, navigation isn't available
+ * listingDetail optional: text rendered next to action buttons, similar style to the tab headers
+ * listingActions optional: buttons that are presented as actions for the expanded item
```

Comment:
do we want to add `propTypes` for `listingDetail` and `listingActions` or do we get rid of the propTypes?